[web-animations] `KeyframeValue` and `AcceleratedEffectKeyframe` should derive from a shared base class

graouts opened the pull request:
#### 893f4b23e527ed71d0d970a9adeb78a111870d00
<pre>
[web-animations] `KeyframeValue` and `AcceleratedEffectKeyframe` should derive from a shared base class
<a href="https://bugs.webkit.org/show_bug.cgi?id=264747">https://bugs.webkit.org/show_bug.cgi?id=264747</a>

Reviewed by Dean Jackson.

As part of the work on threaded animation resolution (see bug 250970), we will need to have the keyframe
interpolation logic currently contained in `KeyframeEffect` also apply to `AcceleratedEffect`. This will
require a fair amount of refactoring to move all the relevant code in a shared base class for both
`KeyframeEffect` and `AcceleratedEffect`.

The first step towards this goal is to have the keyframe class backing those animation effect classes,
`KeyframeValue` and `AcceleratedEffectKeyframe`, use a shared base class. In this patch we introduce a
new `KeyframeInterpolation` class, which is currently empty except for the nested `KeyframeInterpolation::Keyframe`.
It exposes a simple set of methods that `KeyframeValue` and `AcceleratedEffectKeyframe` now override:

    double offset() const;
    std::optional&lt;CompositeOperation&gt; compositeOperation() const;
    bool animatesProperty(Property) const;

We also add a new abstract `AnimatedProperty` which can be either `AnimatableCSSProperty` or `AcceleratedEffectProperty`.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::getKeyframes):
(WebCore::KeyframeEffect::setAnimatedPropertiesInStyle):
(WebCore::KeyframeEffect::computeExtentOfTransformAnimation const):
(WebCore::KeyframeEffect::progressUntilNextStep const):
(WebCore::KeyframeEffect::computeHasImplicitKeyframeForAcceleratedProperty):
* Source/WebCore/animation/KeyframeInterpolation.h: Added.
(WebCore::KeyframeInterpolation::Keyframe::isAcceleratedEffectKeyframe const):
(WebCore::KeyframeInterpolation::Keyframe::isKeyframeValue const):
* Source/WebCore/animation/WebAnimationTypes.h:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::buildObjectForKeyframes):
* Source/WebCore/platform/animation/AcceleratedEffect.cpp:
(WebCore::AcceleratedEffect::Keyframe::Keyframe):
(WebCore::AcceleratedEffect::Keyframe::animatesProperty const):
(WebCore::AcceleratedEffect::Keyframe::clone const):
(WebCore::AcceleratedEffect::create):
(WebCore::AcceleratedEffect::AcceleratedEffect):
(WebCore::AcceleratedEffectKeyframe::clone const): Deleted.
* Source/WebCore/platform/animation/AcceleratedEffect.h:
(WebCore::AcceleratedEffect::keyframes const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::startAnimation):
* Source/WebCore/rendering/style/KeyframeList.cpp:
(WebCore::KeyframeList::operator== const):
(WebCore::KeyframeList::insert):
(WebCore::KeyframeList::hasImplicitKeyframes const):
(WebCore::KeyframeList::copyKeyframes):
(WebCore::KeyframeList::fillImplicitKeyframes):
(WebCore::KeyframeValue::animatesProperty const):
(WebCore::KeyframeValue::containsProperty const): Deleted.
* Source/WebCore/rendering/style/KeyframeList.h:
(WebCore::KeyframeValue::KeyframeValue): Deleted.
(WebCore::KeyframeValue::properties const): Deleted.
(WebCore::KeyframeValue::key const): Deleted.
(WebCore::KeyframeValue::setKey): Deleted.
(WebCore::KeyframeValue::style const): Deleted.
(WebCore::KeyframeValue::setStyle): Deleted.
(WebCore::KeyframeValue::timingFunction const): Deleted.
(WebCore::KeyframeValue::setTimingFunction): Deleted.
(WebCore::KeyframeValue::compositeOperation const): Deleted.
(WebCore::KeyframeValue::setCompositeOperation): Deleted.
(WebCore::KeyframeValue::containsDirectionAwareProperty const): Deleted.
(WebCore::KeyframeValue::setContainsDirectionAwareProperty): Deleted.
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::keyframeStylesForAnimation):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/270648@main">https://commits.webkit.org/270648@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3d8ca7d2aa01ec355fbf3a2ef528a8c30036e1a0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26022 "6 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4634 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27305 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28122 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23835 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26340 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6398 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2063 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23906 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26276 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/3513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/22424 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28702 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/23376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/23749 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/23754 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/27315 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/3164 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/1373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4556 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3624 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3337 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3485 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

grorg commented on Source/WebCore/animation/KeyframeInterpolation.cpp:
```diff
@@ -0,0 +1,31 @@
+/*
```
Why do you need this file?

graouts replied:
I don’t need it at this stage, it’s true. I can add it in the next patch.

graouts committed:
[web-animations] `KeyframeValue` and `AcceleratedEffectKeyframe` should derive from a shared base class
https://bugs.webkit.org/show_bug.cgi?id=264747

Reviewed by Dean Jackson.

As part of the work on threaded animation resolution (see bug 250970), we will need to have the keyframe
interpolation logic currently contained in `KeyframeEffect` also apply to `AcceleratedEffect`. This will
require a fair amount of refactoring to move all the relevant code in a shared base class for both
`KeyframeEffect` and `AcceleratedEffect`.

The first step towards this goal is to have the keyframe class backing those animation effect classes,
`KeyframeValue` and `AcceleratedEffectKeyframe`, use a shared base class. In this patch we introduce a
new `KeyframeInterpolation` class, which is currently empty except for the nested `KeyframeInterpolation::Keyframe`.
It exposes a simple set of methods that `KeyframeValue` and `AcceleratedEffectKeyframe` now override:

    double offset() const;
    std::optional<CompositeOperation> compositeOperation() const;
    bool animatesProperty(Property) const;

We also add a new abstract `AnimatedProperty` which can be either `AnimatableCSSProperty` or `AcceleratedEffectProperty`.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::getKeyframes):
(WebCore::KeyframeEffect::setAnimatedPropertiesInStyle):
(WebCore::KeyframeEffect::computeExtentOfTransformAnimation const):
(WebCore::KeyframeEffect::progressUntilNextStep const):
(WebCore::KeyframeEffect::computeHasImplicitKeyframeForAcceleratedProperty):
* Source/WebCore/animation/KeyframeInterpolation.h: Added.
(WebCore::KeyframeInterpolation::Keyframe::isAcceleratedEffectKeyframe const):
(WebCore::KeyframeInterpolation::Keyframe::isKeyframeValue const):
* Source/WebCore/animation/WebAnimationTypes.h:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::buildObjectForKeyframes):
* Source/WebCore/platform/animation/AcceleratedEffect.cpp:
(WebCore::AcceleratedEffect::Keyframe::Keyframe):
(WebCore::AcceleratedEffect::Keyframe::animatesProperty const):
(WebCore::AcceleratedEffect::Keyframe::clone const):
(WebCore::AcceleratedEffect::create):
(WebCore::AcceleratedEffect::AcceleratedEffect):
(WebCore::AcceleratedEffectKeyframe::clone const): Deleted.
* Source/WebCore/platform/animation/AcceleratedEffect.h:
(WebCore::AcceleratedEffect::keyframes const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::startAnimation):
* Source/WebCore/rendering/style/KeyframeList.cpp:
(WebCore::KeyframeList::operator== const):
(WebCore::KeyframeList::insert):
(WebCore::KeyframeList::hasImplicitKeyframes const):
(WebCore::KeyframeList::copyKeyframes):
(WebCore::KeyframeList::fillImplicitKeyframes):
(WebCore::KeyframeValue::animatesProperty const):
(WebCore::KeyframeValue::containsProperty const): Deleted.
* Source/WebCore/rendering/style/KeyframeList.h:
(WebCore::KeyframeValue::KeyframeValue): Deleted.
(WebCore::KeyframeValue::properties const): Deleted.
(WebCore::KeyframeValue::key const): Deleted.
(WebCore::KeyframeValue::setKey): Deleted.
(WebCore::KeyframeValue::style const): Deleted.
(WebCore::KeyframeValue::setStyle): Deleted.
(WebCore::KeyframeValue::timingFunction const): Deleted.
(WebCore::KeyframeValue::setTimingFunction): Deleted.
(WebCore::KeyframeValue::compositeOperation const): Deleted.
(WebCore::KeyframeValue::setCompositeOperation): Deleted.
(WebCore::KeyframeValue::containsDirectionAwareProperty const): Deleted.
(WebCore::KeyframeValue::setContainsDirectionAwareProperty): Deleted.
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::keyframeStylesForAnimation):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/270648@main
